src/workerd/api/worker-rpc.h

```diff
@@ -96,6 +96,46 @@ class JsRpcClientProvider: public jsg::Object {
       jsg::Lock& js, kj::Vector<kj::StringPtr>& path) = 0;
 };
 
+class JsRpcProperty;
+
+// Represents the promise returned by calling an RPC method. We don't use a regular Promise object,
+// but rather our own custom thenable, so that we can support pipelining on it.
+class JsRpcPromise: public JsRpcClientProvider {
+public:
+  JsRpcPromise(jsg::JsRef<jsg::JsPromise> inner,
+      IoOwn<rpc::JsRpcTarget::CallResults::Pipeline> pipeline)
+      : inner(kj::mv(inner)), pipeline(kj::mv(pipeline)) {}
+
+  rpc::JsRpcTarget::Client getClientForOneCall(
+      jsg::Lock& js, kj::Vector<kj::StringPtr>& path) override;
+
+  // Implement standard Promise interface, especially `then()` so that this works as a custom
+  // thenable.
+  //
+  // Note that we intentionally return jsg::JsValue rather than jsg::JsPromise because we actually
+  // do not want the JSG glue to recognize we're returning a promise triggering behavior that pins
+  // the JsRpcPromise in memory until it resolves. It's actually fine if the JsRpcPromise is GC'd
+  // before the promise resolves.
```

Comment:
> // It's actually fine if the JsRpcPromise is GC'd before the promise resolves.

nit: I'm a bit confused by this, why can the object representing the RPC promise be GC'd before the result is ready? I'm assuming it's because calling `.then()` ensures we aren't dropping the `inner` promise, but in that case it might be helpful to mention that in this comment.

Reply:
I've extended the comments slightly to try to clarify. But essentially the JsRpcPromise is just an API wrapper to provide a JavaScirpt interface, it's not the actual promise. The actual promise keeps running even if the JsRpcPromise goes away.

* e771128b
* ab35cef1